File: java/src/main/java/com/paylater/usecases/UserUseCase.java
package com.paylater.usecases;

import com.paylater.entities.*;
import com.paylater.utils.Misc;
import com.paylater.utils.ResponseValue;
import com.paylater.repository.LocalDatabase;

import java.util.Map;


public class UserUseCase {
    private final LocalDatabase db;

    public UserUseCase(LocalDatabase db) {
        this.db = db;
    }

    /**
     * Create new user
     *
     * @param name        : name of the user
     * @param email       : email id of the user
     * @param creditLimit : credit limit set to the user
     * @return ResponseValue: success-> true if user created successfully otherwise false
     */
    public ResponseValue newUser(String name, String email, int creditLimit) {
        ResponseValue responseValue;
        User user;
        try {
            user = db.getUser(name);
            if (user != null) {
                return ResponseValue.getInstance(false, "User " + name + " Already exists",
                        Misc.getEmptyArrayList());
            }

            user = new User(name, email, creditLimit);
            db.addUser(user);
            return ResponseValue.getInstance(true, user.toString(), Misc.getEmptyArrayList());

        } catch (Exception e) {
            return ResponseValue.getInstance(false, "Error while creating user: " + e.toString(),
                    Misc.getEmptyArrayList());
        }
    }

    /**
     * @param usr    : name of user
     * @param mrchnt : name of merchant
     * @param amt    : amount of transaction
     * @return Map   : success -> true if transaction successful otherwise false
     */
    public ResponseValue transaction(String usr, String mrchnt, String amt) {
        try {
            ResponseValue responseValue;
            int amount = Integer.parseInt(amt);

            // User check
            User user = db.getUser(usr);
            if (user == null) {
                return ResponseValue.getInstance(false, "User " + usr + " not exists",
                        Misc.getEmptyArrayList());
            }

            // Merchant check
            Merchant merchant = db.getMerchant(mrchnt);
            if (merchant == null) {
                return ResponseValue.getInstance(false, "Merchant " + mrchnt + " not exists",
                        Misc.getEmptyArrayList());
            }

            if (user.getCurrentLimit() < amount) {
                responseValue = ResponseValue.getInstance(false, "rejected! (reason: credit limit)",
                        Misc.getEmptyArrayList());
            } else {
                // Update Current Limit
                int currentLimit = user.getCurrentLimit() - amount;
                user.setCurrentLimit(currentLimit);
                db.updateUser(user);

                // Update Discount Received
                double discountReceived = merchant.getDiscountReceived() + (merchant.getDiscountPercentage() * 0.01 * amount);
                merchant.setDiscountReceived(discountReceived);
                db.updateMerchant(merchant);

                responseValue = ResponseValue.getInstance(true, "success!",
                        Misc.getEmptyArrayList());
            }
            return responseValue;
        } catch (Exception e) {
            return ResponseValue.getInstance(false, "Error while executing transaction: " + e.toString(),
                    Misc.getEmptyArrayList());
        }
    }

    /**
     * Payback dues of user
     *
     * @param usr : name of user
     * @param amt : amount to be paid back
     * @return ResponseValue : success-> true if payback successful otherwise false
     */
    public ResponseValue paybackDues(String usr, String amt) {
        ResponseValue responseValue;
        try {
            int amount = Integer.parseInt(amt);
            User user = db.getUser(usr);
            if (user == null) {
                responseValue = ResponseValue.getInstance(false, "User " + usr + " not exists",
                        Misc.getEmptyArrayList());
            } else {
                int currentLimit = user.getCurrentLimit();
                user.setCurrentLimit(currentLimit + amount);

                int dues = user.getCreditLimit() - user.getCurrentLimit();
                if (dues < 0) {
                    responseValue = ResponseValue.getInstance(false,
                            "User " + usr + "is trying to pay more dues", Misc.getEmptyArrayList());
                } else {
                    db.updateUser(user);
                    responseValue = ResponseValue.getInstance(true, usr + "(dues:" + dues + ")",
                            Misc.getEmptyArrayList());
                }
            }
            return responseValue;
        } catch (Exception e) {
            responseValue = ResponseValue.getInstance(false, "Error at payback dues: " + e,
                    Misc.getEmptyArrayList());
            return responseValue;
        }
    }

    /**
     * Reports total dues of users
     *
     * @return ResponseValue: success-> true if list of users with dues otherwise false for exception
     */
    public ResponseValue reportTotalDues() {
        ResponseValue responseValue;
        try {
            Map<String, User> users = db.getUsers();
            int total = 0;

            StringBuilder messageBuilder = new StringBuilder();
            for (Map.Entry<String, User> entry : users.entrySet()) {
                User user = entry.getValue();
                String name = user.getName();
                int dues = user.getCreditLimit() - user.getCurrentLimit();
                if (dues != 0) {
                    total += dues;
                    messageBuilder.append(name).append(":").append(dues).append("\n");
                }
            }
            String message = messageBuilder.toString();

            message += "total: " + total;
            responseValue = ResponseValue.getInstance(true, message, Misc.getEmptyArrayList());

            return responseValue;

        } catch (Exception e) {
            responseValue = ResponseValue.getInstance(false, "Error at report total dues: " + e,
                    Misc.getEmptyArrayList());
            return responseValue;
        }
    }

    /**
     * Report list of users who has `0` current credit limit
     *
     * @return ResponseValue: List of users with `0` current credit limit
     */
    public ResponseValue reportUsersAtCreditLimit() {
        ResponseValue responseValue;
        try {
            User user;
            Map<String, User> users = db.getUsers();
            StringBuilder stringBuilder = new StringBuilder();

            for (Map.Entry<String, User> entry : users.entrySet()) {
                user = entry.getValue();
                if (user.getCurrentLimit() == 0) {
                    stringBuilder.append(user.getName()).append("\n");
                }
            }
            String message = stringBuilder.toString();
            responseValue = ResponseValue.getInstance(true, message, Misc.getEmptyArrayList());
            return responseValue;

        } catch (Exception e) {
            responseValue = ResponseValue.getInstance(false, "Error at report users at credit limit: " + e,
                    Misc.getEmptyArrayList());
            return responseValue;
        }
    }

    /**
     * Report dues of an user
     *
     * @param usr: name of user
     * @return ResponseValue: total dues of an user
     */
    public ResponseValue reportDues(String usr) {
        ResponseValue responseValue;
        try {
            User user = db.getUser(usr);
            if (user != null) {
                int dues = user.getCreditLimit() - user.getCurrentLimit();
                responseValue = ResponseValue.getInstance(true, "dues: (" + dues + ")",
                        Misc.getEmptyArrayList());
            } else {
                responseValue = ResponseValue.getInstance(false, "User " + usr + " not exists",
                        Misc.getEmptyArrayList());
            }
            return responseValue;

        } catch (Exception e) {
            responseValue = ResponseValue.getInstance(false, "Error at report dues: " + e,
                    Misc.getEmptyArrayList());
            return responseValue;
        }
    }
}

File: python/usecases/user_usecase.py
from entities.user import User
from datastore.datastore_interface import DataStoreInterface


class UserUseCase:
    def __init__(self, db: DataStoreInterface):
        self._db = db
        self._user = User

    def new_user(self, name: str, email: str, credit_limit: str) -> dict:
        """
        Create new user
        Args:
            name: name of user
            email: email of user
            credit_limit: credit limit for user

        Returns:
            dict: True if user created otherwise False for exception
        """
        try:
            user = self._db.get_user(name)
            if user:
                result = {'success': False, 'message': f'User "{name}" already exists'}

            else:
                user = self._user(name, email, credit_limit)
                if user:
                    self._db.add_user(user)
                    result = {'success': True, 'message': user}

                else:
                    result = {'success': False, 'message': 'Error while creating user'}

            return result
        except Exception as e:
            return {'success': False, 'message': f'Error while creating user: {e}'}

    def transaction(self, usr: str, mrchnt: str, amt: str) -> dict:
        """
        Carry out transaction for user
        Args:
            usr: name of user
            mrchnt: name of merchant
            amt: amount to be used for transaction

        Returns:
            dict: True if transaction success full otherwise False
        """
        try:
            amount = int(amt)

            user = self._db.get_user(usr)
            if user is False:
                result = {'success': False, 'message': f'User "{usr}" not exists'}
                return result

            merchant = self._db.get_merchant(mrchnt)
            if merchant is False:
                result = {'success': False, 'message': f'Merchant "{mrchnt}" not exists'}
                return result

            if user['current_limit'] < amount:
                result = {'success': False, 'message': 'rejected! (reason: credit limit)'}

            else:
                # Update Current Limit
                user['current_limit'] -= amount
                self._db.update_user(usr, user)

                # Update Discount Received
                merchant['discount_received'] += merchant['discount_percentage'] * 0.01 * amount
                self._db.update_merchant(mrchnt, merchant)

                result = {'success': True, 'message': 'success!'}

            return result

        except Exception as e:
            return {'success': False, 'message': f'Error while executing transaction: {e}'}

    def payback_dues(self, usr: str, amt: str) -> dict:
        """
        Payback dues of user
        Args:
            usr: name user
            amt: amount to be paid back

        Returns:
            dict: True if payback successful otherwise False
        """
        try:
            amount = int(amt)
            user = self._db.get_user(usr)
            if user:
                user['current_limit'] += amount
                dues = user['credit_limit'] - user['current_limit']
                if dues < 0:
                    result = {'success': False, 'message': f'User "{usr}" is trying to paid more dues'}

                else:
                    self._db.update_user(usr, user)
                    result = {'success': True, 'message': f'{usr}(dues:{dues})'}

            else:
                result = {'success': False, 'message': f'User "{usr}" not exists!'}

            return result

        except Exception as e:
            result = {'success': False, 'message': f'Error at payback dues: {e}'}
            return result

    def report_total_dues(self) -> dict:
        """
        Reports total dues of users
        Returns:
            dict: True and list the users with dues otherwise False for exception
        """
        try:
            users = self._db.get_users()
            total = 0
            message = ''
            for user in users:
                name = users[user]['name']
                dues = users[user]['credit_limit'] - users[user]['current_limit']
                if dues != 0:
                    total += dues
                    message += f'{name}: {dues}\n'

            message += f'total: {total}'
            result = {'success': True, 'message': ''}

            return result

        except Exception as e:
            result = {'success': False, 'message': f'Error while getting report of total dues: {e}'}
            return result

    def report_users_at_credit_limit(self) -> dict:
        """
        Report list of users who has `0` current credit limit

        Returns:
            dict: List of users with `0` current credit limit
        """
        try:
            users = self._db.get_users()
            message = ''
            for user in users:
                if users[user]['current_limit'] == 0:
                    message += users[user]['name']

            result = {'success': True, 'message': message}
            return result

        except Exception as e:
            result = {'success': False, 'message': f'Error while getting report of users at credit limit: {e}'}
            return result

    def report_dues(self, usr: str) -> dict:
        """
        Report dues of an user
        Args:
            usr: name of user

        Returns:
            dict: total dues of an user
        """
        try:
            user = self._db.get_user(usr)
            if user:
                dues = user['credit_limit'] - user['current_limit']
                result = {'success': True, 'message': f'dues: {dues}'}

            else:
                result = {'success': False, 'message': f'User "{usr}" not exists!'}

            return result

        except Exception as e:
            result = {'success': False, 'message': f'Error while getting report dues: {e}'}
            return result
